common/rocksdb_env_s3.cpp

```diff
@@ -0,0 +1,432 @@
+                                std::unique_ptr<SequentialFile>* result,
+                                const EnvOptions& options) {
+  assert(s3_util_ != nullptr);
+  result->reset();
+
+  // We read first from local storage and then from S3.
+  auto local_full_path = local_directory_ + fname;
+  auto st = posix_env_->NewSequentialFile(local_full_path, result, options);
+
+  if (!st.ok()) {
+    // If file doesnt exist in local, we copy the file to the local storage from S3
+    auto resp = s3_util_->getObject(s3_key_prefix_ + fname, local_full_path, FLAGS_s3_direct_io);
+    if (!resp.Error().empty()) {
+      LOG(ERROR) << "Error happened when copying the file from S3 to local: "
+                 << resp.Error();
+      return Status::IOError();
+    }
+  }
+
+  // we successfully copied the file, try opening it locally now
+  st = posix_env_->NewSequentialFile(local_full_path, result, options);
+
+  return st;
+}
+
+// open a file for random reading
+Status S3Env::NewRandomAccessFile(const std::string& fname,
+                                  std::unique_ptr<RandomAccessFile>* result,
+                                  const EnvOptions& options) {
+  assert(s3_util_ != nullptr);
+  result->reset();
+
+  // We read first from local storage and then from S3.
+  auto local_full_path = local_directory_ + fname;
+  auto st = posix_env_->NewRandomAccessFile(local_full_path, result, options);
+
+  if (!st.ok()) {
+    // If file doesnt exist in local, we copy the file to the local storage from S3
+    auto resp = s3_util_->getObject(s3_key_prefix_ + fname, local_full_path, FLAGS_s3_direct_io);
+    if (!resp.Error().empty()) {
+      LOG(ERROR) << "Error happened when copying the file from S3 to local: "
+                 << resp.Error();
+      return Status::IOError();
+    }
+  }
+
+  // we successfully copied the file, try opening it locally now
+  st = posix_env_->NewRandomAccessFile(local_full_path, result, options);
+
+  return st;
+}
+
+// S3WritableFile is a wrapper for writing a file in S3. We will perform the
+// operations in a local tmp file and upload to S3 when closing the file.
+class S3WritableFile : public WritableFile {
+ private:
+  const std::string fname_;
+  const std::string s3_fname_;
+  S3Env* env_;
+  std::shared_ptr<common::S3Util> s3_util_;
+  Status status_;
+  std::unique_ptr<WritableFile> local_file_;
+
+ public:
+  S3WritableFile(const std::string& local_fname,
+                 const std::string& s3_fname,
+                 const EnvOptions& options,
+                 S3Env* env,
+                 std::shared_ptr<common::S3Util> s3_util) :
+      fname_(local_fname),
+      s3_fname_(s3_fname),
+      env_(env),
+      s3_util_(s3_util) {
+    auto local_env = env_->GetBaseEnv();
+    auto s = local_env->NewWritableFile(local_fname, &local_file_, options);
+    if (!s.ok()) {
+      LOG(ERROR) << "Error happened when creating writable file in local: " << s.ToString();
+      status_ = s;
```

Comment:
Any reason for setting a status in the constructor? It's odd for the caller to check status as soon as calling the constructor (and the caller might forget to do so). Would it be better to do this in a separate function?

Reply:
I followed the convention in the rocksdb hdfs implementation